core/block_validator.go

```diff
@@ -51,7 +51,7 @@ func NewBlockValidator(config *params.ChainConfig, blockchain *BlockChain, engin
 // validated at this point.
 func (v *BlockValidator) ValidateBody(block *types.Block) error {
 	// Check whether the block's known, and if not, that it's linkable
-	if v.bc.HasBlockAndState(block.Hash(), block.NumberU64()) {
+	if v.bc.HasBlockAndState(block.Hash(), block.NumberU64()) && v.bc.HasSnapshot(block.Hash(), block.NumberU64()) {
```

Comment:
This is not correct, IMO. We should be able to run on trie only, and not require re-execution of blocks just because the snapshots are missing.

Reply:
another way to solve this issue is to delete all the state after the block snapshot root referred to